runtime/src/main/java/org/corfudb/runtime/Backup.java

```diff
@@ -0,0 +1,236 @@
+     */
+    public Backup(String filePath, List<UUID> streamIDs, CorfuRuntime runtime) throws IOException {
+        this.filePath = filePath;
+        this.backupTempDirPath = Files.createTempDirectory(BACKUP_TEMP_DIR_PREFIX).toString();
+        this.streamIDs = streamIDs;
+        this.runtime = runtime;
+        this.timestamp = runtime.getAddressSpaceView().getLogTail();
+    }
+
+    /**
+     * The difference from the constructor above is that this one will automatically discover
+     * tables which requires back up from their tags.
+     *
+     * @param filePath      - the filePath where the generated backup tar file will be placed
+     * @param runtime       - the runtime which is performing the back up
+     * @throws IOException  - when failed to create the temp directory
+     */
+    public Backup(String filePath, CorfuRuntime runtime) throws IOException {
+        this.filePath = filePath;
+        this.backupTempDirPath = Files.createTempDirectory(BACKUP_TEMP_DIR_PREFIX).toString();
+        this.runtime = runtime;
+        this.timestamp = runtime.getAddressSpaceView().getLogTail();
+        this.streamIDs = getTaggedTables();
+    }
+
+    /**
+     * Start the backup process
+     *
+     * @throws IOException
+     */
+    public void start() throws IOException {
+        try {
+            backup();
+            generateTarFile();
+        } catch (IOException e) {
+            log.debug("failed to backup tables: {}", streamIDs);
+            throw e;
+        } finally {
+            cleanup();
+        }
+        log.debug("backup completed");
+    }
+
+    /**
+     * All temp backupTable files will be placed under BACKUP_DIR_PATH directory.
+     *
+     * @throws IOException
+     */
+    private void backup() throws IOException {
+        for (UUID streamId : streamIDs) {
+            String fileName = backupTempDirPath + File.separator + streamId;
+            backupTable(fileName, streamId);
+        }
+        log.debug("successfully backed up {} tables to {} directory", streamIDs.size(), backupTempDirPath);
+    }
+
+    /**
+     * Back up a single table
+     *
+     * If the log is trimmed at timestamp, the backupTable will fail.
+     * If the table has no data to backupTable, it will create a file with empty contents.
+     *
+     * @param fileName   - the name of the backup file
+     * @param uuid       - the uuid of the table which is being backed up
+     * @throws IOException
+     */
+    private void backupTable(String fileName, UUID uuid) throws IOException {
+        try (FileOutputStream fileOutput = new FileOutputStream(fileName)) {
+            StreamOptions options = StreamOptions.builder()
+                    .ignoreTrimmed(false)
+                    .cacheEntries(false)
+                    .build();
+            Stream<OpaqueEntry> stream = (new OpaqueStream(runtime, runtime.getStreamsView().get(uuid, options))).streamUpTo(timestamp);
```

Comment:
Good point. I thinking failing the whole backup would be easier for users to implement some retry mechanism. We don't want users to say backup 6 in 10 tables (because the other 4 failed) and backup the remaining 4 tables in a second run, and then merge the two tar files. I understand your point is that if some tables failed at least we don't lose other successful backups. I think we need some comments from Graham to see how they want to handle this exception.